### ISSUE TITLE
Creating class libraries in .NET Core

### ISSUE THREAD

**Issue by jan-johansson-mr** [dotnet/core#222]
Hi,

By using the (CLI) command **_dotnet new -t lib_**, I've got a skeleton code for a class library, and that's really nice. However, I really don't know how to let another project consume the library that I've created. I've tested to create NuGet package, (CLI) command _**dotnet pack**_, and edit _**NuGet.config**_ file to add a local repository. But it didn't work for me. I've been searching for some documentation to help me, but found very little (but nothing has helped me).

It would be fantastic to get my class libraries to show up in Visual Studio Code when I'm authoring the _**project.json**_ file, but I've not been successful. I really don't know where to turn, so I drop my comments to this forum.

I'm working on Ubuntu 16.04, using Visual Studio Code as my editor.

Kind Regards,
Jan

**Comment by svick** [comment]
Could you describe what did you do and how did it fail?

**Comment by jan-johansson-mr** [comment]
Hi,

First step was to create skeleton code, via _**dotnet new -t lib**_. This creates a class library project template, with a default file name _**Library.cs**_ (my directory, folder, name is _**mylib**_) In this case, I'm trying to understand the tool chain, so next step is simply to build the artifact via _**dotnet restore**_ and _**dotnet build**_.

The produced artifact is default configured for _**netstandard1.6**_.

Now I have an (class library) assembly in the output directory (folder) at _**bin/Debug/netstandard1.6/mylib.dll**_.

This is very nice! But then I want to use my assembly in another project, and it is unclear to me how to do that, so I guessed that maybe the assembly must be packaged into a NuGet package.

To do this, I run the _**dotnet pack**_ command. This produces a NuGet package _**mylib.1.0.0.nupkg**_. Next step is to publish the package (from a local repository), so I created a local directory (in my case I did that at _**.nuget/mypackages/**_, I'm testing this in Ubuntu 16.04).

To make NuGet aware of my repository, I edited the _**NuGet.config**_ file:

_**< configuration >
  < packageSources >
    < add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    < add key="local" value="~/.nuget/mypackages/" />
  < /packageSources >
< /configuration >**_

And then copied _**mylib.1.0.0.nupkg**_ to _**.nuget/mypackages/**_.

I have no idea if I'm doing this right. I've tested to modify the repository path to a canonical path, or to have a simple path (_**mypackages**_) (a simple path is relative to the NuGet root - according to NuGet documentation). But I simply can't use my class library from any project, when I reference the package.

And it does not show up in Visual Studio Code when I edit the _**project.json**_ file.

I need some guidance in how to make class library assemblies available to .NET Core projects, and if it's possible to make it available in the Visual Studio Code.

In short, it fails because I can not use my class libraries in my .NET Core projects.

**Comment by Petermarcu** [comment]
Hi @jan-johansson-mr,

I haven't tried it with creating a NuGet package the way you are trying to do it. I was able to get it working without it by creating this folder structure:

ProjectRoot
|__ src
____|__ app
________|__ Program.cs
________|__ project.json
____|__ library
________|__ Library.cs
________|__ project.json

- Run `dotnet new -t lib` in the library folder and `dotnet new` in the app folder. 

- Update the contents of the apps project.json to reference the library by adding this to the dependencies:

`"library":{
      "version":"1.0.0",
      "target":"project"
    }`

- Update the Program.cs to call something in the library

- Run `dotnet restore` from the library folder.
- Run `dotnet restore` from the app folder.

- Run `dotnet run` from the app folder.

- You should see the result of whatever method you called in the library.

Let me know if this worked or didn't work for you.

**Comment by jan-johansson-mr** [comment]
Hi Peter,

Thanks for your description, it does indeed work!

I want to use assembly (or assemblies) (dll) in my projects. I've not found any clear way to do that (it's easy in VS). However, by following what I wrote earlier today, it does actually work (using dll)! As you know, there is no NuGet utility on native Linux, but as far as I understand it _**dotnet restore**_ does some lifting with the NuGet repository. I found out that once the _**NuGet.config**_ file points to my local repository (where I put the generated NuGet files from _**dotnet pack**_ command), then _**dotnet restore**_ creates the needed directory structure under _**~/.nuget/package**_ directory (folder) and pushes the package(s) there.

Once this is done, it works quite well with both Visual Studio Code, and dotnet CLI.

The only thing that doesn't work that well is intellisense in the _**project.json**_ file. But once the package is referenced in the dependencies "section", with the correct version (no help from intellisense though), it works quite well.

I'm sure more support will come from tooling soon, but until then, it's a little bit of manual labor to get it to work... :-)

**Comment by Petermarcu** [comment]
Yeah, I look forward to the next rev of the tooling. I hope a lot of things will get easier.
